### PR TITLE
Allow cross process communication using custom messages

### LINKED ISSUE #1780: Allow master node to supply data to worker nodes directly
### Is your feature request related to a problem? Please describe.

Pretty much a duplicate of #1506, but it would be extremely beneficial for locust to be able to communicate data between nodes easily. An example that you're probably familiar with is needing to retrieve user accounts from a database for login, and trying to coordinate that between several worker nodes so that the same users are not logged in across multiple workers. 

### Describe the solution you'd like

Adding hooks to allow communication between nodes would be ideal. I'm not sure exactly what the best solution would be but allowing the user to define custom messages and handlers for the greenlets seems like it'd be easy enough to implement and massively improve locust's ability to coordinate itself in distributed mode. 

### Describe alternatives you've considered

The currently accepted workarounds (e.g. mongodb or a REST api) are much less than ideal, especially for something like a docker environment where each additional component must be installed and set up regularly.

### Additional context

I'm available to give this a shot, but I wanted to make sure A) it has value and B) the best implementation from those who are more experienced on this project.

### PR DESCRIPTION
**Pull request by nathan-beam**
This pull fixes #1780 and #1506. 

Added two methods, `register_message` and `send_message` to the two `DistributedRunner` classes.

`register_message` takes a custom message type and a callback function to call when the specified message is received. The callback's parameters are `environment` (obviously the runner's environment) and `msg`, which is the message itself. 

`send_message` takes a custom message type and optional data to send with the message, allowing a user to easily send their own custom messages.

Updated tests/documentation where necessary, added an example implementation to the examples directory (`examples\custom_messages.py`).

### CODE REVIEW

File: locust/runners.py
Comment (cyberw) [inline locust/runners.py:738]: Are you trying to do an f-string? If so I think you are missing the `f` :)

File: examples/custom_messages.py
Comment (cyberw) [inline examples/custom_messages.py:1]: Can we do a more "real world" example? Like the worker asking for some test data (a customer ID or something) and the master replying with it? I think it is one of the most common use cases for this.

File: examples/custom_messages.py
Comment (cyberw) [inline examples/custom_messages.py:29]: Does it make sense for the master to be able to send a message to a specific worker? Sending to all workers is also useful, but sending to a specific one is probably even more so.

**Comment by cyberw** [comment]
Cool stuff. You saw the linting/formatting issues right? I'd like to hold off on this until we have merged and released the new worker distribution style though. So 2.1 at the earliest.

File: examples/custom_messages.py
Comment (nathan-beam) [reply examples/custom_messages.py:1]: Sure, let me add in something a little less theoretical

File: examples/custom_messages.py
Comment (nathan-beam) [reply examples/custom_messages.py:29]: Can you give an example of a use-case where the targeting would be beneficial? In my opinion, since the Master node has so little information with which to target the nodes anyways, it'd be easier to either only register the events you'd want to target on the nodes you're interested in, or include the targeting criteria in the message data somehow, rather than trying to figure out from the master which node you'd want to receive the message.

**Comment by nathan-beam** [comment]
Any ETA on 2.1? I have some functionality that would depend on this, so just want to have a timeline so I can plan things out as best I can. Thanks!

File: examples/custom_messages.py
Comment (nathan-beam) [reply examples/custom_messages.py:29]: Actually after updating the `custom_messages` example I see the usefulness. I'll add that in as well. 👍

**Comment by cyberw** [comment]
It mostly depends on that other PR, and I’m not writing it myself. I’d say maybe a month?

**Review by benesadia2000 — APPROVED** [review]

**Comment by cyberw** [comment]
A small note: please write commit messages in a way that makes sense even if the reader doesnt know what branch/PR they were part of.

**Comment by cyberw** [comment]
Maybe some test cases for when a message is sent but no receiving method is registered? (should log a warning)

**Comment by pappacena** [comment]
+1 for this PR.

I'm facing a similar situation, and I had to mostly copy the `WorkerRunner.worker` method in a subclass. 

I was [preparing a refactoring](https://github.com/locustio/locust/compare/master...pappacena:refactoring-worker-message?expand=1) to make it easier to subclass `WorkerRunner` and add new message types, but this proposal is way better! 

Good job!

**Comment by cyberw** [comment]
@nathan-beam If you add the test cases I mentioned I will merge this and it will become part of the next minor version. No point in waiting for that other PR (and this is not a breaking change anyway).

**Comment by nathan-beam** [comment]
@cyberw Added. Let me know if you think it's missing anything else

**Comment by cyberw** [comment]
Looks pretty good now. 

Just one thing: as most plans are testes in a stand alone environment before moving to distributed, could this be made to work on stand alone (not distributed) runs? (or does it already work that way? If so then just add a test case for it)

**Comment by nathan-beam** [comment]
> Just one thing: as most plans are testes in a stand alone environment before moving to distributed, could this be made to work on stand alone (LocalRunner) runs?

Not sure what "working" would look like on a local runner, since there isn't any messaging infrastructure. Are you alright with just logging a warning? The alternative I can think of is to covertly register a new event and have the `send_message()` method fire that event to emulate sending/receiving a message.

**Comment by cyberw** [comment]
It would be great if we could emulate sending/receiving, so that the same test plan can work standalone and distributed without changes. I havent thought about it too much though, but see if you can't solve it without resorting to any weird hacks (if it is impossible/very hard then lets not do it)

**Comment by nathan-beam** [comment]
Was actually even easier than I was thinking. Just moved the custom messages dictionary and registration the base Runner class and simply called the registered callback from the LocalRunner's `send_message()` function. Should be 100% functional between Master/Worker and Local modes now 👍

File: docs/running-locust-distributed.rst
Comment (cyberw) [inline docs/running-locust-distributed.rst:93]: between master and worker nodes

File: docs/running-locust-distributed.rst
Comment (cyberw) [inline docs/running-locust-distributed.rst:127]: Maybe add an actual User class, so that the example can run for real?

Comment (nathan-beam) [reply docs/running-locust-distributed.rst:127]: Most inline code blocks in the documentation seem to only include the relevant code snippets, I would think to avoid cluttering up the page with familiar concepts. What about adding just a link to the full example in the examples directory?

Comment (cyberw) [reply docs/running-locust-distributed.rst:127]: Sounds good!

**Comment by cyberw** [comment]
Nice work!

**Pull request merged**
